runtime/src/main/scala/akka/grpc/internal/SwitchOnCancel.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import akka.stream.Attributes
+import akka.stream.FanOutShape2
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import akka.util.OptionVal
+
+/**
+ * Identity stage that feeds all incoming events to output 1 until it cancels, then switches over to output 2, after
+ * completing the materialized value future.
+ *
+ * INTERNAL API
+ * @tparam T
+ */
+final private[akka] class SwitchOnCancel[T] extends GraphStage[FanOutShape2[T, T, (Throwable, T)]] {
```

Comment:
Might be useful enough that we should backport it to Akka proper. It's kind of like a Source.recover but in the other direction.

Reply:
What is the status of this PR? Should it be fixed in Akka or here?

Reply:
I think we could keep it as a custom stage here and, maybe decide to port back to Akka as public operator in the future.